src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,396 @@
+	 */
+	aclresult = pg_class_aclcheck(cagg->relid, GetUserId(), ACL_SELECT);
+	aclcheck_error(aclresult, OBJECT_MATVIEW, get_rel_name(cagg->relid));
+	cagg_watermark_cache = cagg_watermark_create(cagg, TopTransactionContext);
+
+	PG_RETURN_INT64(cagg_watermark_cache->value);
+}
+
+static int64
+cagg_compute_watermark(ContinuousAgg *cagg, int64 watermark, bool isnull)
+{
+	if (isnull)
+	{
+		watermark = ts_time_get_min(cagg->partition_type);
+	}
+	else
+	{
+		/*
+		 * The materialized hypertable is already bucketed, which means the
+		 * max is the start of the last bucket. Add one bucket to move to the
+		 * point where the materialized data ends.
+		 */
+		if (ts_continuous_agg_bucket_width_variable(cagg))
+		{
+			/*
+			 * Since `value` is already bucketed, `bucketed = true` flag can
+			 * be added to ts_compute_beginning_of_the_next_bucket_variable() as
+			 * an optimization, if necessary.
+			 */
+			watermark =
+				ts_compute_beginning_of_the_next_bucket_variable(watermark, cagg->bucket_function);
+		}
+		else
+		{
+			watermark = ts_time_saturating_add(watermark,
+											   ts_continuous_agg_bucket_width(cagg),
+											   cagg->partition_type);
+		}
+	}
+
+	return watermark;
+}
+
+TS_FUNCTION_INFO_V1(ts_continuous_agg_watermark_materialized);
+
+/*
+ * Get the materialized watermark for a real-time aggregation query on a
+ * continuous aggregate.
+ *
+ * The difference between this function and `ts_continuous_agg_watermark` is
+ * that this one get the max open dimension of the materialization hypertable
+ * insted of get the stored value in the catalog table.
+ */
+Datum
+ts_continuous_agg_watermark_materialized(PG_FUNCTION_ARGS)
+{
+	const int32 mat_hypertable_id = PG_GETARG_INT32(0);
+	ContinuousAgg *cagg;
+	AclResult aclresult;
+	bool isnull;
+	Hypertable *ht;
+	int64 watermark;
+
+	cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_hypertable_id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_hypertable_id)));
+
+	/*
+	 * Preemptive permission check to ensure the function complains about lack
+	 * of permissions on the cagg rather than the materialized hypertable
+	 */
+	aclresult = pg_class_aclcheck(cagg->relid, GetUserId(), ACL_SELECT);
+	aclcheck_error(aclresult, OBJECT_MATVIEW, get_rel_name(cagg->relid));
+
+	ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
+	watermark = ts_hypertable_get_open_dim_max_value(ht, 0, &isnull);
+
+	watermark = cagg_compute_watermark(cagg, watermark, isnull);
+
+	PG_RETURN_INT64(watermark);
+}
+
+TSDLLEXPORT void
+ts_cagg_watermark_insert(Hypertable *mat_ht, int64 watermark, bool watermark_isnull)
+{
+	ContinuousAgg *cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_ht->fd.id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_ht->fd.id)));
```

Comment:
Since you have already fetched the materialized hypertable, this error would mean that something changed between fetching it and using it. The check here will prevent a crash in that case, but if the above scenario can happen, isn't this an indication of something wrong how the locks are handled?

Reply:
Actually I don't need to populate the ContinuousAgg data structure here, I just need to get the time datatype for the open dimension of the Hypertable passed by parameter.

Fixed!